src/lua/msgpack.c

```diff
@@ -569,8 +616,285 @@ lua_decode_map_header(lua_State *L)
 	return 2;
 }
 
+/**
+ * Allocates a new msgpack object capable of storing msgpack data of the given
+ * size and pushes it to Lua stack. Returns a pointer to the object.
+ */
+static struct luamp_object *
+luamp_new_object(struct lua_State *L, size_t data_len)
+{
+	struct luamp_object *obj = lua_newuserdata(L, sizeof(*obj) + data_len);
```

Comment:
Since this is a pure Lua object, I think it should live on Lua memory so userdata should be fine. We could allocate it somewhere else (on malloc or runtime arena) and expose it to Lua as cdata ptr, but this would look weird, since this kind of object is never used in C alone, only in Lua code. @igormunkin @Totktonada WDYT?

Reply:
Discussed w/ @locker this offline. The best option would be using lightuserdata in case this unpacked msgpack object is allocated somewhere else. Otherwise we have no option other than allocating the memory for it: using malloc requires additional actions with releasing this memory later, but saves Lua memory; userdata is the most convenient way to work with such objects in Lua, but nobody wants to waste the limited GC arena for the data not being used in Lua space. All in all, I'm for the latter approach until we face some performance or memory issues related to this code.